### PR TITLE
Users/changliu2/ignite2024 hotfix random runs

### PR DESCRIPTION
**Pull request by changliu2**
## Purpose

Agents require GPT model output to follow certain format, but LLM models may behave non-deterministically and the code will error out. Errors rarely occur in products.py and researcher.py but they do.
 
* added retry logic to reduce non-determination errors in products.py and researcher.py
* added parsing code to handle certain edge cases in product.py
* handled certain edge cases in BING search results for general information with empty relatedSearchResults.
* added better instruction for output formatting in products.prompty for edge cases
* upgraded all 3 agents to gpt-4o for better parseability and instruction following: https://openai.com/index/introducing-structured-outputs-in-the-api/

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?
**Retries may increase running time up to 3 times.**

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: src/api/agents/researcher/researcher.py [hunk 1]
```diff
@@ -139,6 +148,7 @@ def process(research):
 def research(instructions: str, feedback: str = "No feedback"):
     r = execute(instructions=instructions)
     p = process(r)
+    print(p)
```
Comment (YusakuNo1) [inline src/api/agents/researcher/researcher.py:151]: I think we can keep it clean by removing the print out for success path?

**Review by YusakuNo1 — APPROVED** [review]

File: src/api/agents/researcher/researcher.py
Comment (marlenezw) [inline src/api/agents/researcher/researcher.py:1]: Changes to researcher seem fine as I get the results as expected

File: src/api/agents/product/product.py [hunk 2]
```diff
@@ -84,10 +84,25 @@ def retrieve_products(items, index_name):
 
 
 @trace
-def find_products(context: str) -> Dict[str, any]:
+def find_products(context: str, num_retries: int = 3) -> Dict[str, any]:
```
Comment (marlenezw) [inline src/api/agents/product/product.py:87]: We need to revise this code post Ignite. It sometimes works and sometimes does not. I'm finding it causes the editor to break more times than not. Let's circle back soon

File: src/api/agents/product/product.prompty [hunk 3]
```diff
@@ -40,12 +40,22 @@ queries:
...
+
+You should make sure to use the context the user gives you to generate the queries.
 
 # Output format
 Only output the full array of specialized queries to make to the search index. Limit
 yoursef to 5 queries.
 
 user:
-{{context}}
+context: {{context}}
```
Comment (marlenezw) [inline src/api/agents/product/product.prompty:60]: I think the user input should not be in this format.

File: src/api/agents/designer/designer.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 from openai import AzureOpenAI
 from azure.identity import DefaultAzureCredential, get_bearer_token_provider
 
-def image_gen(image_prompt):
+def image_gen(image_prompt: str):
```
Comment (marlenezw) [inline src/api/agents/designer/designer.py:8]: We are currently not using the designer as dalle 3 availability is restricted

**Pull request closed without merging**